### PR TITLE
Week 4 marker maps A4

### PR DESCRIPTION
**Pull request by mmurray22**
Allows users to add markers to the application and serve them up later. Due to technical difficulties with cloud shell, unable to test yet.

Also, how is the size of this PR? While it encapsulates one complete "feature", is it too big?

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/MapMarkers.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+package com.google.sps.servlets;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.gson.Gson;
+import com.google.gson.sps.data.MapMarker;
```
Comment (ihsan314) [inline portfolio/src/main/java/com/google/sps/servlets/MapMarkers.java]: Did you mean `com.google.sps.data.MapMarker`?

File: portfolio/src/main/java/com/google/sps/servlets/MapMarkers.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+                coordinates.setProperty(MARKER_LAT_COLUMN_NAME, lat);
+                coordinates.setProperty(MARKER_LNG_COLUMN_NAME, lng);
+                datastore.put(coordinates);
+            }
+         response.sendRedirect("/index.html");
+    }
+
+    private static String convertToJson(List<MapMarker> markersToDisplay) {
+        Gson gson = new Gson();
+        String json = gson.toJson(comments);
```
Comment (ihsan314) [inline portfolio/src/main/java/com/google/sps/servlets/MapMarkers.java]: Did you mean `markersToDisplay` instead of `comments`?

**Review by ihsan314 — CHANGES_REQUESTED** [review]
I compiled locally and found the following compile errors, and tried to deduce what you may have meant. In addition to this, make sure that your pom.xml contains the following:
```
    <dependency>
      <groupId>com.google.appengine</groupId>
      <artifactId>appengine-api-1.0-sdk</artifactId>
      <version>1.9.59</version>
    </dependency>
```

File: portfolio/src/main/java/com/google/sps/servlets/MapMarkers.java (hunk 1 above)
Comment (mmurray22) [reply portfolio/src/main/java/com/google/sps/servlets/MapMarkers.java]: yes :) I made that fix but it didn't push for some reason! Will do that in a few moments!

File: portfolio/src/main/java/com/google/sps/servlets/MapMarkers.java (hunk 2 above)
Comment (mmurray22) [reply portfolio/src/main/java/com/google/sps/servlets/MapMarkers.java]: yes! lol I have no idea why this didn't push but I also fixed this locally-will push soon! thank you!

**Comment by ihsan314** [comment]
Also consider using [this](https://gist.github.com/ccondit/55f6760a001a9aa4e2167e81ea9bd73b) to format your Java files appropriately.

**Review by honeybeebaby — APPROVED** [review]
ihsan's comments seem valid, so LGTM

**Review by ihsan314 — APPROVED** [review]

File: portfolio/src/main/java/com/google/sps/data/MapMarker.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.data;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class MapMarker {
```
Comment (kaveh096) [inline portfolio/src/main/java/com/google/sps/data/MapMarker.java]:
```suggestion
public final class MapMarker {
```

File: portfolio/src/main/java/com/google/sps/data/MapMarker.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    private int lat;
+    private int lng;
```
Comment (kaveh096) [inline portfolio/src/main/java/com/google/sps/data/MapMarker.java]:
```suggestion
    private final int lat;
    private final int lng;
```

File: portfolio/src/main/java/com/google/sps/data/MapMarker.java (hunk 3 above)
Comment (kaveh096) [inline portfolio/src/main/java/com/google/sps/data/MapMarker.java]: @ccondit 
No action needed. Just FYI. Internally we use AutoValue library to generate such classes automatically. [It looks like it's open source](https://github.com/google/auto/blob/master/value/userguide/index.md#how-to-use-autovalue) so we could use it for our capstone project. That being said protobuf would have made it even easier for this MapMarker use-case as it wold have take care of wire transportation as well.

File: portfolio/src/main/java/com/google/sps/data/MapMarker.java (hunk 4 above)
Comment (kaveh096) [inline portfolio/src/main/java/com/google/sps/data/MapMarker.java]: I believe you need to use `double` types for both lat and long. That is how maps api would take it.

File: portfolio/src/main/java/com/google/sps/servlets/MapMarkers.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        response.getWriter().println(markerJSON);
+    }
+
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+         String lat_str = request.getParameter("lat");
+         String lng_str = request.getParameter("lng");
+         if ((lat_str != null && !lat_str.isEmpty()) && 
+              (lng_str != null && !lng_str.isEmpty())) 
+            {
```
Comment (kaveh096) [inline portfolio/src/main/java/com/google/sps/servlets/MapMarkers.java]: Style nit
```suggestion
         if (lat_str != null && !lat_str.isEmpty() && lng_str != null && !lng_str.isEmpty()) {
```

File: portfolio/src/main/webapp/index.html [hunk 6]
```diff
@@ -95,6 +95,13 @@ <h1>Michaela Murray's Portfolio</h1>
         <option value="Ramen-Nagi">Ramen Nagi, Palo Alto, California, US</option>
         <option value="Havanna">Havanna, Cuba</option>
     </select>
+    <form action="/markers" type="POST" id="storeMarker">
+        <label>Marker I want to save:</label>
+        <input type="number" id="lat"/>
+        <input type="number" id="lng"/>
```
Comment (kaveh096) [inline portfolio/src/main/webapp/index.html]: Allow decimal input
```suggestion
        <input type="number" step="0.000001" min="-90" max="90" id="lat">
        <input type="number" step="0.000001" min="-180" max="180" id="lng">
```

**Review by kaveh096 — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/data/MapMarker.java (hunk 4 above)
Comment (mmurray22) [reply portfolio/src/main/java/com/google/sps/data/MapMarker.java]: yes indeed! I had it at int for testing purposes but I should update it now :)

File: portfolio/src/main/java/com/google/sps/servlets/fix.sh [hunk 7]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
```
Comment (kaveh096) [inline portfolio/src/main/java/com/google/sps/servlets/fix.sh]: Do you really want this file to be in this PR?

File: portfolio/src/main/java/com/google/sps/data/MapMarker.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+package com.google.sps.data;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class MapMarker {
+    private double lat;
+    private double lng;
```
Comment (ccondit) [inline portfolio/src/main/java/com/google/sps/data/MapMarker.java]: these private members can also be final

File: portfolio/src/main/java/com/google/sps/data/MapMarker.java (hunk 3 above)
Comment (kaveh096) [inline portfolio/src/main/java/com/google/sps/data/MapMarker.java]:
```suggestion
public final class MapMarker {
```

File: portfolio/src/main/java/com/google/sps/data/MapMarker.java (hunk 8 above)
Comment (kaveh096) [inline portfolio/src/main/java/com/google/sps/data/MapMarker.java]:
```suggestion
    private final double lat;
    private final double lng;
```

File: portfolio/src/main/java/com/google/sps/servlets/MapMarkers.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+        response.setContentType("application/json;");
+        response.getWriter().println(markerJSON);
+    }
+
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+         String lat_str = request.getParameter("lat");
+         String lng_str = request.getParameter("lng");
+
+         if (lat_str != null && !lat_str.isEmpty() && lng_str != null && !lng_str.isEmpty()) {
```
Comment (ccondit) [inline portfolio/src/main/java/com/google/sps/servlets/MapMarkers.java]: Formatting looks off here - mind invoking the formatting script?

File: portfolio/src/main/java/com/google/sps/data/MapMarker.java [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+        this.lng = lng;
+    }
+
+    public double getLat() {
+        return lat;
+    }
+
+    public double getLng() {
+        return lng;
+    }
```
Comment (kaveh096) [inline portfolio/src/main/java/com/google/sps/data/MapMarker.java]: No action needed. But i don't think yo're using these two public getter methods yet.

File: portfolio/src/main/java/com/google/sps/servlets/fix.sh [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+# Formats a Java source file using https://github.com/google/google-java-format
+#
+# COMMIT OFTEN AND USE AT YOUR OWN RISK.
+
+JAVA_FORMATTER_JAR=https://github.com/google/google-java-format/releases/download/google-java-format-1.8/google-java-format-1.8-all-deps.jar
+LOCAL_JAVA_FORMATTER_PATH=/tmp/google-java-format-1.8-all-deps.jar
+
+wget -q -nc -O $LOCAL_JAVA_FORMATTER_PATH $JAVA_FORMATTER_JAR
+
+java -jar $LOCAL_JAVA_FORMATTER_PATH --replace $1
```
Comment (ccondit) [inline portfolio/src/main/java/com/google/sps/servlets/fix.sh]: revert this file from the PR

File: portfolio/src/main/webapp/index.html [hunk 12]
```diff
@@ -95,6 +95,13 @@ <h1>Michaela Murray's Portfolio</h1>
         <option value="Ramen-Nagi">Ramen Nagi, Palo Alto, California, US</option>
         <option value="Havanna">Havanna, Cuba</option>
     </select>
+    <form action="/markers" type="POST" id="storeMarker">
+        <label>Marker I want to save:</label>
+        <input type="number" step="0.000001" min="-90" max="90" id="lat">
```
Comment (ccondit) [inline portfolio/src/main/webapp/index.html:100]: can you add labels for each input? it will be good for accessibility.

File: portfolio/src/main/webapp/script.js [hunk 13]
```diff
@@ -67,19 +68,18 @@ function loadMap() {
...
+      desiredLocationLatLng = {lat: 37.445640, lng: -122.160736};
+  }
+  if (document.getElementById("locations").value == "Havanna") {
+      desiredLocationLatLng = {lat: 23.115055, lng: -82.365972};
+  }
+  console.log(desiredLocationLatLng);
+  map = new google.maps.Map(
+      document.getElementById('map'),
+      {center: desiredLocationLatLng, zoom: 16});
+  const marker = new google.maps.Marker({
```
Comment (ccondit) [inline portfolio/src/main/webapp/script.js]: these changes look unrelated - can you revert?

**Review by ccondit — CHANGES_REQUESTED** [review]

File: portfolio/src/main/webapp/script.js [hunk 14]
```diff
@@ -109,3 +109,26 @@ function getComments() {
...
+        for (marker in markers) {
+            var desiredLocationLatLng = {lat: markers.lat, lng: markers.lng};
+            const marker = new google.maps.Marker({
+                position: desiredLocationLatLng, 
+                map: map,
+                animation: google.maps.Animation.DROP,
+                title: 'Saved Pin!'
+            });
+        }
+
```
Comment (kaveh096) [inline portfolio/src/main/webapp/script.js]:
```suggestion
```
nit: extra whitespace

**Review by kaveh096 — APPROVED** [review]

File: portfolio/src/main/java/com/google/sps/data/MapMarker.java (hunk 10 above)
Comment (mmurray22) [reply portfolio/src/main/java/com/google/sps/data/MapMarker.java]: yes indeed I am not using them yet, but they may be useful in the future :)

File: portfolio/src/main/java/com/google/sps/servlets/fix.sh (hunk 7 above)
Comment (mmurray22) [reply portfolio/src/main/java/com/google/sps/servlets/fix.sh]: oops! no indeed not 👍

File: portfolio/src/main/java/com/google/sps/data/MapMarker.java (hunk 8 above)
Comment (mmurray22) [reply portfolio/src/main/java/com/google/sps/data/MapMarker.java]: updated!

File: portfolio/src/main/java/com/google/sps/servlets/MapMarkers.java (hunk 9 above)
Comment (mmurray22) [reply portfolio/src/main/java/com/google/sps/servlets/MapMarkers.java]: will do!

File: portfolio/src/main/java/com/google/sps/servlets/fix.sh (hunk 11 above)
Comment (mmurray22) [reply portfolio/src/main/java/com/google/sps/servlets/fix.sh]: will do!

File: portfolio/src/main/webapp/index.html (hunk 12 above)
Comment (mmurray22) [reply portfolio/src/main/webapp/index.html:100]: will do!

File: portfolio/src/main/webapp/script.js (hunk 13 above)
Comment (mmurray22) [reply portfolio/src/main/webapp/script.js]: reverted!

**Review by ccondit — APPROVED** [review]
